lib/core/client.js

```diff
@@ -1153,6 +1153,36 @@ function write (client, request) {
       .on('drain', onDrain)
       .on('error', onFinished)
       .on('close', onFinished)
+  } else if (util.isIterable(body)) {
+    const iterator = body[Symbol.iterator]()
+    let result = iterator.next()
+    const stringOrBuffer = x => {
+      if (typeof x === 'string') {
+        return 1
+      } else if (util.isBuffer(x)) {
+        return 0
+      } else {
+        throw new Error('Body iterable must yield either string or buffer')
+      }
+    }
+    let chunks = stringOrBuffer(result.value) ? result.value : [result.value]
+    while (true) {
+      if (result.done) break
+      result = iterator.next()
+      if (stringOrBuffer(result.value)) {
+        chunks += result.value
+      } else {
+        chunks.push(result.value)
+      }
+    }
```

Comment:
Close, I would probably do it like this:

```js
async function write(iterable, socket) {
  try {
    if (!contentLength) {
      // chunked transfer
    }
    for await (const chunk of iterable) {
      socket.write(chunk)
    }
    socket.write('\r\n', 'ascii')

    request.body = null
  } catch (err) {
    util.destroy(socket, err)
  } 
}
```

The iterable could be huge and not appropriate to read into memory.

Reply:
You basically need to do the same thing as for streams.

Reply:
Ahh okay I'll try this out. Thank you!

Reply:
making some progress here, but since this is taking me quite a bit of time i don't want to impede the overall progress on the library. @ronag would you be interested in maybe doing a pair programming call and working through this together? No worries if you're too busy right now; i'll keep plugging away at it

Reply:
This week is swamped for me. If you are still working on it next week we could try to do some pair programming.

Reply:
Okay sounds good!